### PR TITLE
feat(connector): support gzip compression for file source

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
This PR support (automatically) read files compressed in gzip format, and currently only gzip compression is supported.
- Users can define `compression_format` in the create table statement, eg. `compression_format='gzip'` or `compression_format='gz'`, this will read all files with the .gz suffix.
- If  `compression_format` is not defined or `compression_format='None'` , file reader will automatically read and decompress .gz and .gzip files.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

This pr add a new field in file source connector, which is to read the compressed file. We need to add corresponding changes in https://docs.risingwave.com/docs/current/ingest-from-gcs/
The specific behavior is: 
- Users can define `compression_format` in the create table statement, eg. `compression_format='gzip'` or `compression_format='gz'`, this will read all files with the .gz suffix.
- If  `compression_format` is not defined or `compression_format='None'` , file reader will automatically read and decompress .gz and .gzip files.

 @hengm3467 for visibility.

</details>

### CODE REVIEW

File: src/connector/src/source/filesystem/opendal_source/opendal_reader.rs [hunk 1]
```diff
@@ -134,7 +142,20 @@ impl<Src: OpendalSource> OpendalReader<Src> {
         let stream_reader = StreamReader::new(
             reader.map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e)),
         );
-        let buf_reader = BufReader::new(stream_reader);
+
+        let buf_reader: Pin<Box<dyn AsyncRead + Send>> = match decompression_format {
```
[the remark this answers is not in the document]
Comment (hzxa21) [reply src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:131]: +1

File: src/connector/src/source/filesystem/opendal_source/mod.rs [hunk 2]
```diff
@@ -53,6 +53,9 @@ pub struct GcsProperties {
 
     #[serde(flatten)]
     pub unknown_fields: HashMap<String, String>,
+
+    #[serde(rename = "decompression_format")]
+    pub decompression_format: Option<String>,
```
Comment (hzxa21) [inline src/connector/src/source/filesystem/opendal_source/mod.rs:58]: Shall we make `decompression_format` a enum? See https://github.com/risingwavelabs/risingwave/blob/267e6c9410c13a23fe064c00ae46d9163e7de30e/src/common/src/config.rs#L558 as an example.

File: src/connector/src/source/filesystem/s3/mod.rs [hunk 3]
```diff
@@ -41,6 +41,8 @@ pub struct S3PropertiesCommon {
     pub secret: Option<String>,
     #[serde(rename = "s3.endpoint_url")]
     pub endpoint_url: Option<String>,
+    #[serde(rename = "decompression_format")]
+    pub decompression_format: Option<String>,
```
Comment (hzxa21) [inline src/connector/src/source/filesystem/s3/mod.rs:45]: ditto

**Review by hzxa21 — COMMENTED** [review]
Rest LGTM

File: src/connector/src/source/filesystem/opendal_source/opendal_reader.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:131]: Yes, current implementation is already like this.

File: src/connector/src/source/filesystem/opendal_source/mod.rs (hunk 2 above)
Comment (wcy-fdu) [reply src/connector/src/source/filesystem/opendal_source/mod.rs:58]: SGTM

Comment (wcy-fdu) [reply src/connector/src/source/filesystem/opendal_source/mod.rs:58]: Oh a bit hard, `WithOptions` can only be derived for structs.

File: src/connector/with_options_source.yaml [hunk 4]
```diff
@@ -33,6 +33,9 @@ GcsProperties:
     field_type: String
     required: false
     default: Default::default
+  - name: decompression_format
```
Comment (xxchan) [inline src/connector/with_options_source.yaml:36]: Should this be called `compression_format` instead...?

File: e2e_test/s3/fs_source_v2.py [hunk 5]
```diff
@@ -29,6 +29,12 @@ def format_json(data):
         for file in data
     ]
 
+def format_json_gzip(data):
+    return [
+        '\n'.join([json.dumps(item) for item in file])
+        for file in data
+    ]
```
Comment (xxchan) [inline e2e_test/s3/fs_source_v2.py:36]: This seems unused

Comment (xxchan) [reply e2e_test/s3/fs_source_v2.py:36]: ... or could we merge these test scripts?

File: src/connector/Cargo.toml [hunk 6]
```diff
@@ -23,6 +23,7 @@ arrow-schema = { workspace = true }
 arrow-schema-iceberg = { workspace = true }
 arrow-select = { workspace = true }
 assert_matches = "1"
+async-compression = "0.4.5"
```
Comment (wangrunji0408) [inline src/connector/Cargo.toml:26]: Seems that the "tokio" feature gate should be enabled.

```suggestion
async-compression = { version = "0.4.5", features = ["tokio"] }
```

**Comment by wcy-fdu** [comment]
@hzxa21 PTAL❤️

File: src/connector/src/source/filesystem/file_common.rs [hunk 7]
```diff
@@ -141,3 +142,11 @@ pub struct FsPageItem {
 }
 
 pub type FsPage = Vec<FsPageItem>;
+
+#[derive(Debug, Clone, PartialEq, Display, Deserialize)]
+pub enum DecompressionFormat {
```
Comment (hzxa21) [inline src/connector/src/source/filesystem/file_common.rs:147]: By convention, I think it is better to name this `CompressionFormat`. This also aligns with the variable name used in other places of this PR.

File: src/connector/src/source/filesystem/opendal_source/mod.rs [hunk 8]
```diff
@@ -53,6 +54,9 @@ pub struct GcsProperties {
 
     #[serde(flatten)]
     pub unknown_fields: HashMap<String, String>,
+
+    #[serde(rename = "compression_format")]
+    pub compression_format: Option<DecompressionFormat>,
```
Comment (hzxa21) [inline src/connector/src/source/filesystem/opendal_source/mod.rs:59]: Given that we have `None` compression format defined in the enum, I think we don't need the `Option` here. We can just  let the `compression_format` default to `CompressionFormat::None` if not specified.

File: src/connector/src/source/filesystem/opendal_source/opendal_reader.rs [hunk 9]
```diff
@@ -117,7 +128,27 @@ impl<Src: OpendalSource> OpendalReader<Src> {
...
+                    let gzip_decoder = GzipDecoder::new(stream_reader);
+                    Box::pin(BufReader::new(gzip_decoder)) as Pin<Box<dyn AsyncRead + Send>>
+                } else {
+                    Box::pin(BufReader::new(stream_reader)) as Pin<Box<dyn AsyncRead + Send>>
+                }
+            }
+            Some(format) => {
+                let error_message = format!("The input compression format '{}' is not supported. Currently, only gzip format compression is supported.", format);
+                return Err(io::Error::new(io::ErrorKind::InvalidInput, error_message).into());
+            }
```
Comment (hzxa21) [inline src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:149]: This arm is not needed if my comment above is adopted.

File: src/connector/src/source/filesystem/opendal_source/opendal_reader.rs [hunk 10]
```diff
@@ -130,11 +129,13 @@ impl<Src: OpendalSource> OpendalReader<Src> {
         );
 
         let buf_reader: Pin<Box<dyn AsyncRead + Send>> = match compression_format {
-            Some(DecompressionFormat::Gzip) => {
+            CompressionFormat::Gzip => {
+                println!("这里11object_name = {:?}", object_name);
```
Comment (hzxa21) [inline src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:133]: remove?

File: src/connector/src/source/filesystem/opendal_source/opendal_reader.rs [hunk 11]
```diff
@@ -130,11 +129,13 @@ impl<Src: OpendalSource> OpendalReader<Src> {
...
                 let gzip_decoder = GzipDecoder::new(stream_reader);
                 Box::pin(BufReader::new(gzip_decoder)) as Pin<Box<dyn AsyncRead + Send>>
             }
-            None => {
+            CompressionFormat::None => {
+                println!("这里object_name = {:?}", object_name);
```
Comment (hzxa21) [inline src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:138]: remove?

**Review by hzxa21 — APPROVED** [review]
LGTM.

- The PR description is outdated. Please update it accordingly.
- This is a user-facing feature. Please update the release note as well.

**Comment by wcy-fdu** [comment]
@hengm3467  This pr add a new field in gcs connector, and we need to apply corresponding changes in our doc, the detail is in release note. Let me know if you have any questions, thanks~

**Pull request merged**